crates/ide_assists/src/handlers/generate_default_from_new.rs

```diff
@@ -0,0 +1,259 @@
+use crate::{
+    assist_context::{AssistContext, Assists},
+    utils, AssistId,
+};
+use syntax::{
+    ast::{self, Adt, Impl, NameOwner},
+    AstNode, Direction,
+};
+use test_utils::mark;
+
+// Assist: generate_default_from_new
+//
+// Generates default implementation from new method.
+//
+// ```
+// struct Example { _inner: () }
+//
+// impl Example {
+//     pub fn n$0ew() -> Self {
+//         Self { _inner: () }
+//     }
+// }
+// ```
+// ->
+// ```
+// struct Example { _inner: () }
+//
+// impl Example {
+//     pub fn new() -> Self {
+//         Self { _inner: () }
+//     }
+// }
+//
+// impl Default for Example {
+//     fn default() -> Self {
+//         Self::new()
+//     }
+// }
+// ```
+pub(crate) fn generate_default_from_new(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let fn_node = ctx.find_node_at_offset::<ast::Fn>()?;
+    let fn_name = fn_node.name()?;
+
+    if fn_name.text() != "new" {
+        mark::hit!(other_function_than_new);
+        return None;
+    }
+
+    if fn_node.param_list()?.params().next().is_some() {
+        mark::hit!(new_function_with_parameters);
+        return None;
+    }
+
+    let impl_ = fn_node.syntax().ancestors().into_iter().find_map(ast::Impl::cast)?;
+
+    let insert_location = impl_.syntax().text_range();
+
+    acc.add(
+        AssistId("generate_default_from_new", crate::AssistKind::Generate),
+        "Generate a Default impl from a new fn",
+        insert_location,
+        move |builder| {
+            let default_fn_syntax = default_fn_node_for_new(impl_);
+            if let Some(code) = default_fn_syntax {
+                builder.insert(insert_location.end(), code)
+            }
+        },
+    )
+}
+
+fn default_fn_node_for_new(impl_: Impl) -> Option<String> {
+    // the code string is this way due to formatting reason
+    let code = r#"    fn default() -> Self {
+        Self::new()
+    }"#;
+    let struct_name = impl_.self_ty()?.syntax().to_string();
+    let struct_ = impl_
+        .syntax()
+        .siblings(Direction::Prev)
+        .filter_map(ast::Struct::cast)
+        .find(|struct_| struct_.name().unwrap().text() == struct_name)?;
+
+    let adt = Adt::cast(struct_.syntax().clone())?;
```

Comment:
Ah this won't work if the definition is in a different file. But I have to apologize I gave you the wrong pointer here, what we actually want to do is just copy the impl block of our `new` function and change it to use implement the `Default` trait. So `utils::generate_trait_impl_text` wont help.

roughly speaking what you wanna do is
```rust
format!(
	"impl {} for {} {{
	fn default() -> Self {{
		Self::new()
	}}	
}}",
	  impl_.trait_()?,
	  impl_.self_ty()?
);
"""

Reply:
no problem. I will quickly update the implementation.